[6X_STABLE] Set replication slot's restart_lsn to not greater than redo location

l-wang opened the pull request:
This is a backport of https://github.com/greenplum-db/gpdb/pull/14710

PS: The following changes were made for tests to be compatible with 6X_STABLE: 
(I have mentioned the same in the commit message, but pasting here again for reviewers' convenience. 

    PS: The following changes were made for tests to be compatible with 6X_STABLE:
    1. s/pg_switch_wal()/pg_switch_xlog()/
    2. Added SELECT pg_sleep() after pg_ctl_start() to reduce flakiness
    
    The following changes are specifically for Test 5 which we added in 7X for
    additional test coverage for WAL retention in general:
    3. Removed commands that set GUC wal_recycle because 6X_STABLE doesn't have it
    and this test in 6X may not need it.  So far this test continues to succeed in
    6X even without this GUC and I have not observed flakiness. One reason could be
    that 6X caculates number of WALs to recycle differently from 7X. In the future
    if test 5 becomes flaky, the lack of the wal_recycle GUC could be the potential
    root cause and we should consider either backporting it or writing the test
    differently.
    4. s/pg_ls_waldir()/pg_ls_dir('./pg_xlog')/
    5. Replaced pg_control_checkpoint() with a helper UDF pg_controldata_redo_lsn()
    because 6X_STABLE does not have a built-in equivalent.

soumyadeep2007 commented on src/test/isolation2/sql/pg_rewind_fail_missing_xlog.sql:
```diff
@@ -0,0 +1,297 @@
+-- Start again. Start from a clean shutdown does not perform CHECKPOINT
+2: SELECT pg_ctl_start(datadir, port) from gp_segment_configuration where role = 'm' and content = 0;
+-- Wait for a few seconds for the mirror to be ready to accept connections
+2: SELECT pg_sleep(5);
```
We use the following pattern in other tests to deterministically wait for mirror promotion:

```sql
-- stop a primary in order to trigger a mirror promotion
SELECT pg_ctl((SELECT datadir FROM gp_segment_configuration WHERE role = 'p' AND content = 1), 'stop');

-- trigger failover
select gp_request_fts_probe_scan();

-- wait for content 1 (earlier mirror, now primary) to finish the promotion
0U: SELECT 1;
```

l-wang replied:
Right. However in this particular case we are doing pg_ctl `start` rather than `stop` for a segment in the mirror role, and we don't expect a failover. So can use the same pattern as `stop`. I dislike what I'm doing here either, but didn't find an easier way.

l-wang replied:
`Select 1` wouldn't work either. It wouldn't block because the mirror is marked down.

soumyadeep2007 reviewed (commented):
>3. Removed commands that set GUC wal_recycle because 6X_STABLE doesn't have it.
If Test 5 becomes flaky in the future this could be responsible and we should
consider writing the test differently or remove Test 5 all together.

I am leaning towards backporting the wal_recycle GUC (and marking it hidden) from https://github.com/postgres/postgres/commit/475861b2615dd63ae8431d811749a6f9a15bbfd6 (excluding the wal_init_zero stuff, it should be 3-4 lines) and using it here.

One may argue that it might help on certain filesystems such as ZFS, so why not just enable it?! No reason why we can't. We can mark it visible to users later on too. No need to be blocked on that decision for the time being.

> Removed 3 queries that use pg_control_checkpoint() because 6X_STABLE doesn't
have an equivalent.

How did you replace it?
You could write a small plpy UDF that just runs  `pg_controldata` and greps the checkpoint from that, if you needed to.

l-wang commented:
> > 5\. Removed commands that set GUC wal_recycle because 6X_STABLE doesn't have it.
> >     If Test 5 becomes flaky in the future this could be responsible and we should
> >     consider writing the test differently or remove Test 5 all together.
> 
> I am leaning towards backporting the wal_recycle GUC (and marking it hidden) from [postgres/postgres@475861b](https://github.com/postgres/postgres/commit/475861b2615dd63ae8431d811749a6f9a15bbfd6) (excluding the wal_init_zero stuff, it should be 3-4 lines) and using it here.
> 
> One may argue that it might help on certain filesystems such as ZFS, so why not just enable it?! No reason why we can't. We can mark it visible to users later on too. No need to be blocked on that decision for the time being.

In hindsight as I'm backporting this commit I think converting these tests into tap tests are more in line with upstream test style and I'm thinking of doing that in a separate PR that starts from the main branch. I hesitate backporting a GUC just for a test that may later be rewritten. If we want this GUC to help certain filesystems that should definitely be a separate PR.

> > Removed 3 queries that use pg_control_checkpoint() because 6X_STABLE doesn't
> > have an equivalent.
> 
> How did you replace it? You could write a small plpy UDF that just runs `pg_controldata` and greps the checkpoint from that, if you needed to.

I only removed them and did not replace them, because without them the test still serves its purpose. The purpose of this test is to verify if old wal get truncated during checkpoint.

In 7X we verifies the following:
- Replication slot's restart_lsn should now advance to the checkpoint's redo location. -- nice to have
- Replication slot's restart_lsn should advance exactly to the checkpoint's redo location. -- nice to have
- And some WAL file(s) should be removed. -- this is the goal

In 6X we verifies the following:
- Replication slot's restart_lsn should now advance. -- nice to have
- And some WAL file(s) should be removed. -- this is the goal

And again this test itself is not even testing the fix (the fix is to retain wals more conservatively which is tested by test 1-4) but really adding additional test coverage for wal removal in general. In man branch because we have more helper functions and GUCs it was not so difficult to write it as an isolation2 test, but still I think a tap test would be more suitable. @soumyadeep2007 Let me know what you think. I'm very glad you asked and I'd be happy to add an UDF to grep checkpoint in pg_controldata if you still think that's necessary for 6X.

soumyadeep2007 commented:
> In hindsight as I'm backporting this commit I think converting these tests into tap tests are more in line with upstream test style and I'm thinking of doing that in a separate PR that starts from the main branch. I hesitate backporting a GUC just for a test that may later be rewritten. If we want this GUC to help certain filesystems that should definitely be a separate PR.

I think it's okay since we have put in a lot of effort into this. Yes, a TAP test would have been ideal but this isolation2 test should serve nicely (we have many tests like this which could be TAP). It's important coverage we should add now rather than later.

I am on board with getting the `wal_recycle` backport in as a separate PR. This PR can wait on that one.

l-wang commented:
> > In hindsight as I'm backporting this commit I think converting these tests into tap tests are more in line with upstream test style and I'm thinking of doing that in a separate PR that starts from the main branch. I hesitate backporting a GUC just for a test that may later be rewritten. If we want this GUC to help certain filesystems that should definitely be a separate PR.
> 
> I think it's okay since we have put in a lot of effort into this. Yes, a TAP test would have been ideal but this isolation2 test should serve nicely (we have many tests like this which could be TAP). It's important coverage we should add now rather than later.
> 
> I am on board with getting the `wal_recycle` backport in as a separate PR. This PR can wait on that one.

Alternatively, I could just take this test out all together and replace it with a different negative test for this fix. (Read: don't count number of wals in pg_xlog directory, instead only make sure restart_lsn advances). That way we still have a negative test and avoid the potential flakiness this test may introduce due to lack of the `wal_recycle` GUC. 

The original intention of this test in 7X was two folds:
1. A negative test for the fix
2. Add wal removal test coverage in general. 

For 6X, I would like to get this fix in while we figure out what's the best way to test wal removal.

l-wang commented:
@soumyadeep2007 I added the UDF to precisely locate checkpoint redo position as you suggested. 
@soumyadeep2007 @ashwinstar I did not backport `wal_recycle` GUC because both in my local environment and in CI I did not obverse a single test failure due to the lack of this GUC. I think the reason is that 6X calculates number of WALs to recycle differently from 7X, and in this particular test we managed to still remove WALs even when wal_recycle is enabled. I updated the commit message to explicitly record the concern of potential flakiness for future reference just in case tests evolve. For now I do not think we need to introduce a fix for a problem that does not exist yet.

ashwinstar commented:
> @soumyadeep2007 I added the UDF to precisely locate checkpoint redo position as you suggested. 
> 
> @soumyadeep2007 @ashwinstar I did not backport `wal_recycle` GUC because both in my local environment and in CI I did not obverse a single test failure due to the lack of this GUC. I think the reason is that 6X calculates number of WALs to recycle differently from 7X, and in this particular test we managed to still remove WALs even when wal_recycle is enabled. I updated the commit message to explicitly record the concern of potential flakiness for future reference just in case tests evolve. For now I do not think we need to introduce a fix for a problem that does not exist yet.

Yes based on offline discussion, sounds reasonable to me. Lets proceed with what we have and worry only if we have to in future.

soumyadeep2007 reviewed (approved):
LGTM! Thanks for taking this through!

l-wang commented:
Thank you @soumyadeep2007 @ashwinstar for reviewing!
